core/metacling/src/TClingDataMemberInfo.cxx

```diff
@@ -352,10 +352,9 @@ Longptr_t TClingDataMemberInfo::Offset()
       //   static constexpr Long64_t something = std::numeric_limits<Long64_t>::max();
       cling::Interpreter::PushTransactionRAII RAII(fInterp);
 
-      if (Longptr_t addr = reinterpret_cast<Longptr_t>(fInterp->getAddressOfGlobal(GlobalDecl(VD))))
-         return addr;
-      auto evalStmt = VD->ensureEvaluatedStmt();
-      if (evalStmt && evalStmt->Value) {
+      // We can't reassign constexpr or const variables. We can compute the
+      // initializer.
+      if (VD->hasInit() && (VD->isConstexpr() || VD->getType().isConstQualified())) {
```

Comment:
I think this would hugely benefit from being its own commit, for logical reasons and also for future `git-bisect` runs.

Reply:
Agreed, that's the entire commit with the test included.

Reply:
No, it's still mudged together with the changes in `AutoloadLibraryMU`. If I understand correctly, they are two separate changes, no?

Reply:
No, the changes in the `AutoloadLibraryMU` enable the tests in the unittests, otherwise there is not really a clear way how to test that change there.

Reply:
I still don't understand... In any case, having this in two commits will improve bisectability which is a clear win in my eyes.

Reply:
FWIW to make the test pass, I need only the changes in `AutoloadLibraryGenerator` and `AutoloadLibraryMU` , not this one. So, why this change and what does it have to do with autoloading?

Reply:
That change was necessary on my machine. In both cases it makes sense that whenever we stop autoloading to disable the callback-based symbol resolution of the JIT as that's what autoloading is intended to limit.

Reply:
I maintain that this change related to initialization of variables and `constexpr` handling. Please note that I'm not commenting on the two checks in `AutoloadLibraryGenerator` and `AutoloadLibraryMU`; while I don't understand them, I acknowledge that they are needed. This change in `core/metacling/src/TClingDataMemberInfo.cxx` does not seem needed.

Reply:
I guess I am lost. This check here avoids going to the expensive `getAddressOfGlobal` operation for `const` and `constexpr` variable declarations. Instead, we evaluate the initializer as we know it is constant and it is faster. The test changes similar to:

```cpp
 gInterpreter->Declare("static const double var4 = 4.2;");
 EXPECT_DOUBLE_EQ(4.2, *(double*)gROOT->GetGlobal("var4")->GetAddress());

 // Make sure ROOT's Core constexpr constants work
 EXPECT_EQ(3000, *(int*)gROOT->GetGlobal("kError")->GetAddress());
```

test that branch here where we get the constant initializer of the variable compute it and return it as a value as required by the interface. In a similar vain:

```cpp
 TGlobal *GeoManagerInfo = gROOT->GetGlobal("gGeoManager");
 TInterpreter::SuspendAutoLoadingRAII autoloadOff(gInterpreter);
 EXPECT_TRUE(-1L == (ptrdiff_t)GeoManagerInfo->GetAddress());
```

tests the fact that when we suspend autoloading we do not resolve `gGeoManager` through the slow path by loading the library and doing dlsym.

Reply:
I'm not debating the usefulness of the check or whether it's correct or not. I am complaining that it has *nothing* to do with autoloading, is explained *nowhere* in the commit message nor PR summary, and is *not* required to pass the added test (according to my setup). IMHO this means it *must* be in its own commit with a *proper* explanation, or maybe even in an entirely separate PR.

And I fail to see how your second comment regarding the test involving `gGeoManager` has anything to do with the code I am commenting on in `core/metacling/src/TClingDataMemberInfo.cxx`?

Reply:
In the commit message says: `In addition when computing the offsets of a constant variable
declaration we compute the initializers instead of searching in the shared objects. `

Please suggest wording of the commit message that makes you happy.

> And I fail to see how your second comment regarding the test involving gGeoManager has anything to do with the code I am commenting on in core/metacling/src/TClingDataMemberInfo.cxx?

We've moved:

```cpp
if (Longptr_t addr = reinterpret_cast<Longptr_t (fInterp->getAddressOfGlobal(GlobalDecl(VD))))
         return addr;
```
in terms of priorities - so for me it is quite connected even though it's not on that particular line, it has been downgraded in terms of priority...

Reply:
Okay, I think I responded to that and my reasoning is that’s one feature in my opinion. I fail to see your arguments why I should separate these changes since they are entangled. First we delay calling the slow getSymbolAddress operation and at the same time enable the computation of the constexpr variables from the initializer list.

Reply:
To avoid us circling around forever, here is the concrete organization that I have in mind: https://github.com/hahnjo/root/commits/reduce-symbol-search/ As far as I can tell, all tests pass in the intermediate stages. The commit messages are still quite empty, could probably be enhanced. And it's debatable whether the second and third commit should actually be split at this level; but as I expected it seems to be just important that the slow path is only moved after evaluating the initializer of constant variables...

Reply:
Does that sequence works if you revert the workarounds first?

Either way, I don't want to bikeshed more on how to structure commits. I still think these are details that are a matter or preference of the people who do the work. For better or worse ROOT has not have even a policy that tests go with the same commit.

If you feel strongly about that ordering you are welcome to take over the PR. I am not quite sure if that's the way we should review first responders' time-critical fixes, though.

Reply:
> Does that sequence works if you revert the workarounds first?

Do you mean moving the last commit first to test the `constexpr`? I can try tomorrow...

> Either way, I don't want to bikeshed more on how to structure commits. I still think these are details that are a matter or preference of the people who do the work. For better or worse ROOT has not have even a policy that tests go with the same commit.

Maybe it's close to bikeshedding and I'm known to be very pedantic regarding commit organization, but in a core critical area such as symbol search, I think every minute we can invest now will potentially save us hours at some point in the future. I pointed out this request from the very beginning, which I think is a very normal code review procedure in functioning open source projects. The response was that it's technically not possible because the changes are entangled which I provided proof that it's not the case. I personally don't see any reason *not* to take the better organization.

> If you feel strongly about that ordering you are welcome to take over the PR. I am not quite sure if that's the way we should review first responders' time-critical fixes, though.

I'm sorry if you feel that way but I will restate that it is my strong personal belief that code review in open source projects should work like this. For me this is not a standard that should be sacrificed in the name of time-criticality (even though that is the reason I tried to follow up as quickly as possible).

Reply:
> > Does that sequence works if you revert the workarounds first?
>
> Do you mean moving the last commit first to test the `constexpr`? I can try tomorrow...

Yes, it also works. So you want me to push the re-organized commits to this PR?

Reply:
Sure.